pkg/decoder/delimiter.go

```diff
@@ -0,0 +1,145 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2017 Datadog, Inc.
+
+package decoder
+
+import (
+	"regexp"
+	"sync"
+	"time"
+)
+
+// Line represents content separated by two '\n'
+type Line struct {
+	content []byte
+}
+
+// NewLine returns a new Line
+func NewLine(content []byte) *Line {
+	return &Line{
+		content: content,
+	}
+}
+
+// Delimiter
+type Delimiter interface {
+	LineChan() chan *Line
+}
+
+// SingleLineDelimiter reads lines from lineChan and uses lineBuffer to send them
+type SingleLineDelimiter struct {
+	lineChan   chan *Line
+	lineBuffer *LineBuffer
+}
+
+// NewSingleLineDelimiter returns a new NewSingleLineDelimiter
+func NewSingleLineDelimiter(outputChan chan *Output) *SingleLineDelimiter {
+	lineChan := make(chan *Line)
+	lineBuffer := NewLineBuffer(outputChan)
+	delimiter := SingleLineDelimiter{
+		lineChan:   lineChan,
+		lineBuffer: lineBuffer,
+	}
+	go delimiter.run()
+	return &delimiter
+}
+
+// LineChan returns lineChan
+func (d *SingleLineDelimiter) LineChan() chan *Line {
+	return d.lineChan
+}
+
+// run reads lines from lineChan to process them
+func (d *SingleLineDelimiter) run() {
+	for line := range d.lineChan {
+		d.process(line)
+	}
+}
+
+// process sends or truncates lines using lineBuffer
+func (d *SingleLineDelimiter) process(line *Line) {
+	lineLen := len(line.content)
+	if lineLen < contentLenLimit {
+		// send line
+		d.lineBuffer.Add(line)
+		d.lineBuffer.Flush()
+	} else {
+		// truncate content
+		d.lineBuffer.AddAndTruncate(line)
+	}
+}
+
+// outdatedDuration represents the time we want to wait before sending content stored in lineBuffer
+// when mo more line is received
+const outdatedDuration = 1 * time.Second
+
+// MultiLineDelimiter reads lines from lineChan and uses lineBuffer to send them when a new line matches with re or the timer is fired
+type MultiLineDelimiter struct {
+	lineChan   chan *Line
+	lineBuffer *LineBuffer
+	re         *regexp.Regexp
+	timer      *time.Timer
+	mu         sync.Mutex
+}
+
+// NewMultiLineDelimiter returns a new MultiLineDelimiter
+func NewMultiLineDelimiter(outputChan chan *Output, re *regexp.Regexp) *MultiLineDelimiter {
+	lineChan := make(chan *Line)
+	lineBuffer := NewLineBuffer(outputChan)
+	timer := time.NewTimer(outdatedDuration)
+	delimiter := MultiLineDelimiter{
+		lineChan:   lineChan,
+		lineBuffer: lineBuffer,
+		re:         re,
+		timer:      timer,
+	}
+	go delimiter.run()
+	return &delimiter
+}
+
+// LineChan returns lineChan
+func (d *MultiLineDelimiter) LineChan() chan *Line {
```

Comment:
Do we need the `LineChan` methods? As opposed to make `lineChan` public and avoid code duplication?

Reply:
We need it in `decoder` and since we can't expose variables in `interface`, I need this function but @mnshdw already told me that the `Decoder` interface was not really useful.

We could have two references in `decoder` instead, a first one to `SingleLineDecoder` and another one to `MultiLineDecoder`.

The last solution would be to change `Delimiter` as a `struct` and abstract something else. WDYT ?

Reply:
Ok I see, then I think current behavior is fine